upgrade-2.x.sh

```diff
@@ -1250,6 +1282,43 @@ FETCHED_SLUG=$(CURL_CA_BUNDLE="${TMPCRT}" ${CURL} -H "Authorization: Bearer ${AP
 SLUG=${FORCED_SLUG:-$FETCHED_SLUG}
 HOST_OS_VERSION=${META_BALENA_VERSION:-${VERSION_ID}}
 
+# Must query for target version and target image, which includes registry endpoint,
+# if started from app UUID.
+if [ -n "$app_uuid" ]; then
+    # Retrieve release and target image information for the provided app UUID
+    # and release commit. Verify that the release succeeded and has not been invalidated.
+    # We expect at most a single release. Catch command failure for handling below.
+    _query_res=$(CURL_CA_BUNDLE="${TMPCRT}" ${CURL} \
+        -H "Content-Type: application/json" -H "Authorization: Bearer ${APIKEY}" \
+        "${API_ENDPOINT}/v6/release?\$select=raw_version&\$expand=contains__image(\$expand=image(\$select=is_stored_at__image_location,content_hash))&\$filter=commit%20eq%20%27${release_commit}%27%20and%20(belongs_to__application/any(bta:bta/is_host%20and%20bta/uuid%20eq%20%27${app_uuid}%27))%20and%20status%20eq%20'success'%20and%20is_invalidated%20eq%20false" || echo "fail")
+
+    # Verify the result includes a json row
+    _has_row=$(echo "${_query_res}" | jq -r ".d[]" || echo "")
+    if [ -z "$_has_row" ]; then
+        if [ "${_query_res}" = "fail" ]; then
+            log ERROR "Target release query failed"
+        else
+            log ERROR "Target release not found or not valid"
+        fi
+    fi
+
+    target_version=$(echo "${_query_res}" | jq -r ".d[] | .raw_version")
+
+    # Be defensive with this output since it's more complex
+    target_image=$(echo "${_query_res}" | jq -r ".d[] | .contains__image[0].image[0] | [.is_stored_at__image_location, .content_hash] | \"\(.[0])@\(.[1])\"" || echo "")
```

Comment:
I'm wondering if this could also be a problem with a potential registry name change, is this image the source of truth or should it be the image passed via the target state? @thgreasi ?

Reply:
@kb2ma I realize that you are trying to optimize the number of API requests here and fetch both the version & image url in one go, but at this point it might be premature optimization.
I think you would appreciate keeping the two HUP flows (proxy vs supervisor triggered) use as much common flow as possible, so for the mvp I would personally just do the minimum changes needed, and rely on the ` target_image=$(get_image_location "${target_version}")` that's a bit later for finding the image location (if not already provided by the supervisor).